test/NuGet.Server.V2.Tests/NuGetODataControllerTests.cs

```diff
@@ -294,6 +314,646 @@ public void FindPackagesByIdDoesNotReturnDeletedPackages()
+            [InlineData("A|B", "1.0|")]
+            public void GetUpdatesReturnsEmptyResultsIfInputIsMalformed(string id, string version)
+            {
+                // Arrange
+                var repo = Mock.Of<IServerPackageRepository>();
+                var v2Service = new TestableNuGetODataController(repo);
+                v2Service.Request = new HttpRequestMessage(HttpMethod.Get, "https://localhost:8081/");
+
+                // Act
+                var result = v2Service.GetUpdates(
+                    new ODataQueryOptions<ODataPackage>(new ODataQueryContext(NuGetV2WebApiEnabler.BuildNuGetODataModel(), typeof(ODataPackage)), v2Service.Request),
+                    id,
+                    version,
+                    includePrerelease: false,
+                    includeAllVersions: true,
+                    targetFrameworks: null,
+                    versionConstraints: null)
+                    .ExpectOkNegotiatedContentResult<IQueryable<ODataPackage>>();
+
+                // Assert
+                Assert.Empty(result);
+            }
+
+            [Fact]
+            public void GetUpdatesIgnoresItemsWithMalformedVersions()
+            {
+                // Arrange
+                var repo = new Mock<IServerPackageRepository>(MockBehavior.Strict);
+                repo.Setup(r => r.GetPackages()).Returns(
+                    new[]
+                {
+                        CreatePackageWithDefaults("Foo", "1.0.0", listed: true),
+                        CreatePackageWithDefaults("Foo", "1.1.0", listed: true),
+                        CreatePackageWithDefaults("Foo", "1.2.0-alpha", listed: true),
+                        CreatePackageWithDefaults("Foo", "1.2.0", listed: true),
+                        CreatePackageWithDefaults("Qux", "2.0", listed: true) ,
+                }.AsQueryable());
+                var v2Service = new TestableNuGetODataController(repo.Object);
+                v2Service.Request = new HttpRequestMessage(HttpMethod.Get, "https://localhost:8081/");
+
+                // Act
+                var result = v2Service.GetUpdates(
+                    new ODataQueryOptions<ODataPackage>(new ODataQueryContext(NuGetV2WebApiEnabler.BuildNuGetODataModel(), typeof(ODataPackage)), v2Service.Request),
+                    "Foo|Qux",
+                    "1.0.0|abcd",
+                    includePrerelease: false,
+                    includeAllVersions: false,
+                    targetFrameworks: null,
+                    versionConstraints: null)
+                    .ExpectQueryResult<ODataPackage>()
+                    .GetInnerResult()
+                    .ExpectOkNegotiatedContentResult<IQueryable<ODataPackage>>();
+
+                // Assert
+                Assert.Equal(1, result.Count());
+                AssertPackage(new { Id = "Foo", Version = "1.2.0" }, result.First());
+            }
+
+            [Fact]
+            public void GetUpdatesReturnsVersionsNewerThanListedVersion()
```

Comment:
The implementation in the NuGetODataController does little more than parse arguments and forward to repository.GetUpdatesCore().

This test therefore fails since GetUpdatesCore is not mocked. And mocking GetUpdatesCore kind of destroys the point of this test.